src/exception/types.ts

```diff
@@ -19,23 +19,59 @@
 
 export type ObjectValues<T> = T[keyof T];
 
+// base
 export type ExceptionRecord = {
+  program_name: string;
   schema: string;
   requested_core_field: string;
   requested_exception_value: string;
 };
-export type ProgramExceptionRecord = {
-  program_name: string;
-} & ExceptionRecord;
+
+// program exception
+export type ProgramExceptionRecord = ExceptionRecord;
 
 // type after validation
 export type ProgramException = {
   programId: string;
-  exceptions: {
-    schema: string;
-    coreField: string;
-    exceptionValue: ExceptionValueType;
-  }[];
+  exceptions: ReadonlyArray<ProgramExceptionRecord>;
+};
+
+// Entity
+export type BaseEntityExceptionRecord = ExceptionRecord & { submitter_donor_id: string };
+
+export type SpecimenExceptionRecord = BaseEntityExceptionRecord & {
+  submitter_specimen_id: string;
+};
+
+export type FollowUpExceptionRecord = BaseEntityExceptionRecord & {
+  submitter_followup_id: string;
+};
+
+export type EntityExceptionRecord = SpecimenExceptionRecord | FollowUpExceptionRecord;
+export type EntityExceptionRecords = (SpecimenExceptionRecord | FollowUpExceptionRecord)[];
+export type ExceptionRecords =
+  | ReadonlyArray<ProgramExceptionRecord>
+  | ReadonlyArray<SpecimenExceptionRecord>
+  | ReadonlyArray<FollowUpExceptionRecord>;
+
+/**
+ * entity values to be valid EntityException exceptions arrays
+ * provide typing to tsv schema string to exception schema string
+ */
+export const EntityValues: Record<string, Exclude<keyof EntityException, 'programId'>> = {
+  specimen: 'specimen',
+  follow_up: 'follow_up',
+} as const;
```

Comment:
this is another case in which I'd have used an enum instead, but I remember @joneubank having strong opinions against it (I think?)